### ISSUE TITLE
Imagenet Weights

### ISSUE THREAD

**Issue by ilkarman** [kylemin/S3D#2]
Thanks for providing the pytorch version of this model! To get the 72.2% result on Kinetics, you must have started from imagenet-trained weights? Would you mind sharing those weights (or a conversion script perhaps on i3d-rgb weights; to map them to your model)?

At the moment it seems I would need to write a custom script that loads the kernel(1,1,1) weights as-is and then splits the kernel(3,3,3) weights [using](https://github.com/piergiaj/pytorch-i3d/blob/master/models/rgb_imagenet.pt)

**Comment by ilkarman** [comment]
I had a go at creating my own conversion from the i3d weights:

```
    def load_state_dict(self, path):

        target_weights = torch.load(path)
        own_state = self.state_dict()

        for name, param in target_weights.items():
            if '.bn.' not in name:
                try:
                    s3d_name = i3d_to_s3d[name]
                    if param.size()[-1] != 1:
                        param = torch.mean(param, 2, keepdim=True)
                    else:
                        param = param.data
                    own_state[s3d_name].copy_(param)
                    print('Copied param: {}'.format(s3d_name))
                except KeyError:
                    print('No data for param: {}'.format(name))
```

Using this custom dict:

```
i3d_to_s3d = {
    # STEM
    'Conv3d_1a_7x7.conv3d.weight': 'base.0.conv.weight',
    'Conv3d_2b_1x1.conv3d.weight': 'base.2.conv.weight',
    'Conv3d_2c_3x3.conv3d.weight': 'base.3.conv.weight',
    # Mixed_3b
    'Mixed_3b.b0.conv3d.weight':  'base.5.branch0.0.conv.weight',
    'Mixed_3b.b1a.conv3d.weight': 'base.5.branch1.0.conv.weight',
    'Mixed_3b.b1b.conv3d.weight': 'base.5.branch1.1.conv.weight',
    'Mixed_3b.b2a.conv3d.weight': 'base.5.branch2.0.conv.weight',
    'Mixed_3b.b2b.conv3d.weight': 'base.5.branch2.1.conv.weight',
    'Mixed_3b.b3b.conv3d.weight': 'base.5.branch3.1.conv.weight',
    # Mixed_3c
    'Mixed_3c.b0.conv3d.weight':  'base.6.branch0.0.conv.weight',
    'Mixed_3c.b1a.conv3d.weight': 'base.6.branch1.0.conv.weight',
    'Mixed_3c.b1b.conv3d.weight': 'base.6.branch1.1.conv.weight',
    'Mixed_3c.b2a.conv3d.weight': 'base.6.branch2.0.conv.weight',
    'Mixed_3c.b2b.conv3d.weight': 'base.6.branch2.1.conv.weight',
    'Mixed_3c.b3b.conv3d.weight': 'base.6.branch3.1.conv.weight',
    # Mixed_4b
    'Mixed_4b.b0.conv3d.weight':  'base.8.branch0.0.conv.weight',
    'Mixed_4b.b1a.conv3d.weight': 'base.8.branch1.0.conv.weight',
    'Mixed_4b.b1b.conv3d.weight': 'base.8.branch1.1.conv.weight',
    'Mixed_4b.b2a.conv3d.weight': 'base.8.branch2.0.conv.weight',
    'Mixed_4b.b2b.conv3d.weight': 'base.8.branch2.1.conv.weight',
    'Mixed_4b.b3b.conv3d.weight': 'base.8.branch3.1.conv.weight',
    # Mixed_4c
    'Mixed_4c.b0.conv3d.weight':  'base.9.branch0.0.conv.weight',
    'Mixed_4c.b1a.conv3d.weight': 'base.9.branch1.0.conv.weight',
    'Mixed_4c.b1b.conv3d.weight': 'base.9.branch1.1.conv.weight',
    'Mixed_4c.b2a.conv3d.weight': 'base.9.branch2.0.conv.weight',
    'Mixed_4c.b2b.conv3d.weight': 'base.9.branch2.1.conv.weight',
    'Mixed_4c.b3b.conv3d.weight': 'base.9.branch3.1.conv.weight',
    # Mixed_4d
    'Mixed_4d.b0.conv3d.weight':  'base.10.branch0.0.conv.weight',
    'Mixed_4d.b1a.conv3d.weight': 'base.10.branch1.0.conv.weight',
    'Mixed_4d.b1b.conv3d.weight': 'base.10.branch1.1.conv.weight',
    'Mixed_4d.b2a.conv3d.weight': 'base.10.branch2.0.conv.weight',
    'Mixed_4d.b2b.conv3d.weight': 'base.10.branch2.1.conv.weight',
    'Mixed_4d.b3b.conv3d.weight': 'base.10.branch3.1.conv.weight',
    # Mixed_4e
    'Mixed_4e.b0.conv3d.weight':  'base.11.branch0.0.conv.weight',
    'Mixed_4e.b1a.conv3d.weight': 'base.11.branch1.0.conv.weight',
    'Mixed_4e.b1b.conv3d.weight': 'base.11.branch1.1.conv.weight',
    'Mixed_4e.b2a.conv3d.weight': 'base.11.branch2.0.conv.weight',
    'Mixed_4e.b2b.conv3d.weight': 'base.11.branch2.1.conv.weight',
    'Mixed_4e.b3b.conv3d.weight': 'base.11.branch3.1.conv.weight',
    # Mixed_4f
    'Mixed_4f.b0.conv3d.weight':  'base.12.branch0.0.conv.weight',
    'Mixed_4f.b1a.conv3d.weight': 'base.12.branch1.0.conv.weight',
    'Mixed_4f.b1b.conv3d.weight': 'base.12.branch1.1.conv.weight',
    'Mixed_4f.b2a.conv3d.weight': 'base.12.branch2.0.conv.weight',
    'Mixed_4f.b2b.conv3d.weight': 'base.12.branch2.1.conv.weight',
    'Mixed_4f.b3b.conv3d.weight': 'base.12.branch3.1.conv.weight',
    # Mixed_5b
    'Mixed_5b.b0.conv3d.weight':  'base.14.branch0.0.conv.weight',
    'Mixed_5b.b1a.conv3d.weight': 'base.14.branch1.0.conv.weight',
    'Mixed_5b.b1b.conv3d.weight': 'base.14.branch1.1.conv.weight',
    'Mixed_5b.b2a.conv3d.weight': 'base.14.branch2.0.conv.weight',
    'Mixed_5b.b2b.conv3d.weight': 'base.14.branch2.1.conv.weight',
    'Mixed_5b.b3b.conv3d.weight': 'base.14.branch3.1.conv.weight',
    # Mixed_5c
    'Mixed_5c.b0.conv3d.weight':  'base.15.branch0.0.conv.weight',
    'Mixed_5c.b1a.conv3d.weight': 'base.15.branch1.0.conv.weight',
    'Mixed_5c.b1b.conv3d.weight': 'base.15.branch1.1.conv.weight',
    'Mixed_5c.b2a.conv3d.weight': 'base.15.branch2.0.conv.weight',
    'Mixed_5c.b2b.conv3d.weight': 'base.15.branch2.1.conv.weight',
    'Mixed_5c.b3b.conv3d.weight': 'base.15.branch3.1.conv.weight',
}
```

Seems to load the weights okay (ignoring logits and bn). For ST the 3x3x3 are averaged into (1,3,3); (3,1,1) is ignored due to mismatch in channels.

**Comment by ilkarman** [comment]
Sorry I realised that the link contains the ported tensorflow weights which are kinetics + imagenet; not just imagenet.

I think then that I need to use: `models.googlenet(pretrained=True).state_dict()` instead like so:

```
    def load_state_dict(self):

        target_weights = models.googlenet(pretrained=True).state_dict()
        own_state = self.state_dict()

        for name, param in target_weights.items():
            if '.bn.' not in name:
                try:
                    s3d_name = inception_to_s3d[name]
                    param = param.unsqueeze(2)  # Add temporal-dimension (always 1)
                    own_state[s3d_name].copy_(param)
                    print('Copied param: {}'.format(s3d_name))
                except KeyError:
                    print('No data for param: {}'.format(name))
```

The branch names don't quite match up but they do match in terms of param.size()?

```
inception_to_s3d = {
    # STEM
    'conv1.conv.weight': 'base.0.conv.weight',
    'conv2.conv.weight': 'base.2.conv.weight',
    'conv3.conv.weight': 'base.3.conv.weight',
    # 3a
    'inception3a.branch1.conv.weight':   'base.5.branch0.0.conv.weight',
    'inception3a.branch2.0.conv.weight': 'base.5.branch1.0.conv.weight',
    'inception3a.branch2.1.conv.weight': 'base.5.branch1.1.conv.weight',
    'inception3a.branch3.0.conv.weight': 'base.5.branch2.0.conv.weight',
    'inception3a.branch3.1.conv.weight': 'base.5.branch2.1.conv.weight',
    'inception3a.branch4.1.conv.weight': 'base.5.branch3.1.conv.weight',
    # 3b
    'inception3b.branch1.conv.weight':   'base.6.branch0.0.conv.weight',
    'inception3b.branch2.0.conv.weight': 'base.6.branch1.0.conv.weight',
    'inception3b.branch2.1.conv.weight': 'base.6.branch1.1.conv.weight',
    'inception3b.branch3.0.conv.weight': 'base.6.branch2.0.conv.weight',
    'inception3b.branch3.1.conv.weight': 'base.6.branch2.1.conv.weight',
    'inception3b.branch4.1.conv.weight': 'base.6.branch3.1.conv.weight',
    # 4a
    'inception4a.branch1.conv.weight':   'base.8.branch0.0.conv.weight',
    'inception4a.branch2.0.conv.weight': 'base.8.branch1.0.conv.weight',
    'inception4a.branch2.1.conv.weight': 'base.8.branch1.1.conv.weight',
    'inception4a.branch3.0.conv.weight': 'base.8.branch2.0.conv.weight',
    'inception4a.branch3.1.conv.weight': 'base.8.branch2.1.conv.weight',
    'inception4a.branch4.1.conv.weight': 'base.8.branch3.1.conv.weight',
    # 4b
    'inception4b.branch1.conv.weight':   'base.9.branch0.0.conv.weight',
    'inception4b.branch2.0.conv.weight': 'base.9.branch1.0.conv.weight',
    'inception4b.branch2.1.conv.weight': 'base.9.branch1.1.conv.weight',
    'inception4b.branch3.0.conv.weight': 'base.9.branch2.0.conv.weight',
    'inception4b.branch3.1.conv.weight': 'base.9.branch2.1.conv.weight',
    'inception4b.branch4.1.conv.weight': 'base.9.branch3.1.conv.weight',
    # 4c
    'inception4c.branch1.conv.weight':   'base.10.branch0.0.conv.weight',
    'inception4c.branch2.0.conv.weight': 'base.10.branch1.0.conv.weight',
    'inception4c.branch2.1.conv.weight': 'base.10.branch1.1.conv.weight',
    'inception4c.branch3.0.conv.weight': 'base.10.branch2.0.conv.weight',
    'inception4c.branch3.1.conv.weight': 'base.10.branch2.1.conv.weight',
    'inception4c.branch4.1.conv.weight': 'base.10.branch3.1.conv.weight',
    # 4d
    'inception4d.branch1.conv.weight':   'base.11.branch0.0.conv.weight',
    'inception4d.branch2.0.conv.weight': 'base.11.branch1.0.conv.weight',
    'inception4d.branch2.1.conv.weight': 'base.11.branch1.1.conv.weight',
    'inception4d.branch3.0.conv.weight': 'base.11.branch2.0.conv.weight',
    'inception4d.branch3.1.conv.weight': 'base.11.branch2.1.conv.weight',
    'inception4d.branch4.1.conv.weight': 'base.11.branch3.1.conv.weight',
    # 4e
    'inception4e.branch1.conv.weight':   'base.12.branch0.0.conv.weight',
    'inception4e.branch2.0.conv.weight': 'base.12.branch1.0.conv.weight',
    'inception4e.branch2.1.conv.weight': 'base.12.branch1.1.conv.weight',
    'inception4e.branch3.0.conv.weight': 'base.12.branch2.0.conv.weight',
    'inception4e.branch3.1.conv.weight': 'base.12.branch2.1.conv.weight',
    'inception4e.branch4.1.conv.weight': 'base.12.branch3.1.conv.weight',
    # 5a
    'inception5a.branch1.conv.weight':   'base.14.branch0.0.conv.weight',
    'inception5a.branch2.0.conv.weight': 'base.14.branch1.0.conv.weight',
    'inception5a.branch2.1.conv.weight': 'base.14.branch1.1.conv.weight',
    'inception5a.branch3.0.conv.weight': 'base.14.branch2.0.conv.weight',
    'inception5a.branch3.1.conv.weight': 'base.14.branch2.1.conv.weight',
    'inception5a.branch4.1.conv.weight': 'base.14.branch3.1.conv.weight',
    # 5b
    'inception5b.branch1.conv.weight':   'base.15.branch0.0.conv.weight',
    'inception5b.branch2.0.conv.weight': 'base.15.branch1.0.conv.weight',
    'inception5b.branch2.1.conv.weight': 'base.15.branch1.1.conv.weight',
    'inception5b.branch3.0.conv.weight': 'base.15.branch2.0.conv.weight',
    'inception5b.branch3.1.conv.weight': 'base.15.branch2.1.conv.weight',
    'inception5b.branch4.1.conv.weight': 'base.15.branch3.1.conv.weight',
}
```

**Comment by kylemin** [comment]
Hi,
I have no idea why I have not received any notifications. I hope it is not too late.
I think these are what you were looking for?
[i3d-weights-pt](https://drive.google.com/drive/folders/195sbKGT7FNKabQlQ3hGNE3oGQCFeLDBT?usp=sharing)

**Comment by ilkarman** [comment]
Thank you @kylemin !
